File: 백준/Gold/14500. 테트로미노/테트로미노.py
N, M = map(int, input().split())
board = []
visit = [[0 for _ in range(M)] for __ in range(N)]
dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]
for _ in range(N):
    board.append(list(map(int, input().split())))

answer = 0


def out_of_range(x, y):
    return x < 0 or x >= N or y < 0 or y >= M


def dfs(cx, cy, visit, depth, local_sum):
    global answer
    if depth == 4:
        answer = max(answer, local_sum)
        return
    for i in range(4):
        nx, ny = cx + dx[i], cy + dy[i]
        if out_of_range(nx, ny) or visit[nx][ny]:
            continue
        visit[nx][ny] = 1
        dfs(nx, ny, visit, depth + 1, local_sum + board[nx][ny])
        visit[nx][ny] = 0


for i in range(N):
    for j in range(M):
        visit[i][j] = 1
        dfs(i, j, visit, 1, board[i][j])
        visit[i][j] = 0

        for p in range(4):
            cnt = 1
            local_sum = board[i][j]

            for q in range(4):
                nx, ny = i + dx[q], j + dy[q]
                if p == q or out_of_range(nx, ny):
                    continue
                cnt += 1
                local_sum += board[nx][ny]

            if cnt == 4:
                answer = max(answer, local_sum)

print(answer)

File: 백준/Gold/14500. 테트로미노/테트로미노.java
import java.util.Scanner;

public class Main
{
    static int N, M;
    static int[][] arr;
    static int ans;
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static int[][] vis;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        N = sc.nextInt();
        M = sc.nextInt();

        arr = new int[N][M];
        vis = new int[N][M];

        for (int i=0; i<N; i++){
            for (int j=0; j<M; j++){
                arr[i][j] = sc.nextInt();
            }
        }
        for (int i=0; i<N; i++) {
            for (int j = 0; j < M; j++) {
                vis[i][j] = 1;
                dfs(i, j, 0, arr[i][j]);
                vis[i][j] = 0;
            }
        }
        System.out.print(ans);
    }
    public static void dfs(int cx,int cy, int depth, int sum){
        if (depth == 3){
            ans = Math.max(ans, sum);
            return;
        }
        for (int i=0; i<5; i++){
            if (i < 4){

                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (outOfRange(nx, ny)){
                    continue;
                }
                if (vis[nx][ny] == 1){
                    continue;
                }
                vis[nx][ny] = 1;
                dfs(nx, ny, depth+1, sum+arr[nx][ny]);
                vis[nx][ny] = 0;
            }
            else{
                if (depth == 1){
                    for (int j=0; j<4; j++){
                        int nx = cx + dx[j];
                        int ny = cy + dy[j];
                        if (outOfRange(nx, ny))continue;
                        if (vis[nx][ny]==1) continue;
                        for (int k=0; k<4;k++){
                            if (j==k)continue;
                            int nnx = cx + dx[k];
                            int nny = cy + dy[k];
                            if (outOfRange(nnx, nny))continue;
                            if(vis[nnx][nny]==1)continue;
                            vis[nx][ny] = 1;
                            vis[nnx][nny] = 1;
                            dfs(nnx, nny, depth+2,sum+arr[nx][ny]+arr[nnx][nny]);
                            vis[nx][ny] = 0;
                            vis[nnx][nny] = 0;
                        }
                    }
                }
            }
        }
    }

    private static boolean outOfRange(int nx, int ny) {
        return nx < 0 || nx >= N || ny < 0 || ny >= M;
    }
}
